### ISSUE TITLE
[BUG] Cython created frame objects have an incorrect value for `f_back`

### ISSUE THREAD

**Issue by pablogsal** [cython/cython#4207]
**Describe the bug**
The frame objects passed to profile or trace functions have incorrect values of `f_back` attribute

**To Reproduce**
Code to reproduce the behaviour:
## mod.pyx file
```cython
# cython: profile=True
# cython: trace=True

def foo(callback, *args):
    bar(callback, args)

def bar(callback, args):
    callback(*args)
```

#main.py file

```python
import sys
from mod import foo

def trace(frame, *args):
    print(frame.f_code.co_name, frame.f_back.f_code.co_name if frame.f_back else None)

sys.setprofile(trace)

def foop():
    ...

def barp():
    ...

foo(foop)
```

this prints:

```
<module> None
foo <module>
bar <module>
foop <module>
foop <module>
bar <module>
foo <module>
<module> None
<module> None
```

Notice that this is incorrect as this claims that the parent frame of every cython and python frame is `<module>`, which is not correct.

**Expected behavior**

This example should print:

```
```
<module> None
foo <module>
bar foo
foop bar
foop bar
bar foo
foo <module>
<module> None
<module> None
```
```

This has to be achieved by correctly populating `f_back` when Cython creates frames that are passed to the trace function.

**Comment by scoder** [comment]
Seems something we should do in `__Pyx_TraceSetupAndCall()`.
https://github.com/cython/cython/blob/2d8145b1f7a21e7e68b23cbccdc2335ff6ef65e9/Cython/Utility/Profile.c#L266

**Comment by pablogsal** [comment]
> Seems something we should do in `__Pyx_TraceSetupAndCall()`.

Yeah, I think there is some small subtlety with respect to reuse the frames within Cython.

**Comment by pablogsal** [comment]
Also, notice that this needs to be fixed after creation because in CPython we automatically link against the previous frame in the interpreter state, which is never a Cython one.
